extensions/markdown/src/extension.ts

```diff
@@ -0,0 +1,170 @@
+}
+
+function getMarkdownUri(document: vscode.TextDocument) {
+	return document.uri.with({ scheme: 'markdown', query: document.uri.toString() });
+}
+
+function openPreview(sideBySide?: boolean): void {
+	const activeEditor = vscode.window.activeTextEditor;
+	if (!activeEditor) {
+		vscode.commands.executeCommand('workbench.action.navigateBack');
+		return;
+	}
+
+	let markdownPreviewUri = getMarkdownUri(activeEditor.document);
+	vscode.commands.executeCommand('vscode.previewHtml',
+		markdownPreviewUri,
+		getViewColumn(sideBySide),
+		`Preview '${path.basename(activeEditor.document.fileName)}'`);
+}
+
+function getViewColumn(sideBySide): ViewColumn {
+	const active = vscode.window.activeTextEditor;
+	if (!active) {
+		return ViewColumn.One;
+	}
+
+	if (!sideBySide) {
+		return active.viewColumn;
+	}
+
+	switch (active.viewColumn) {
+		case ViewColumn.One:
+			return ViewColumn.Two;
+		case ViewColumn.Two:
+			return ViewColumn.Three;
+	}
+
+	return active.viewColumn;
+}
+
+
+class MDDocumentContentProvider implements TextDocumentContentProvider {
+	private _context: ExtensionContext;
+	private _onDidChange = new EventEmitter<Uri>();
+
+	constructor(context: ExtensionContext) {
+		this._context = context;
+	}
+
+	private getMediaPath(mediaFile) {
+		return this._context.asAbsolutePath(path.join('media', mediaFile));
+	}
+
+	private fixHref(resource: Uri, href: string) {
+		if (href) {
+			// Return early if href is already a URL
+			if (Uri.parse(href).scheme) {
+				return href;
+			}
+			// Otherwise convert to a file URI by joining the href with the resource location
+			return Uri.file(path.join(path.dirname(resource.fsPath), href)).toString();
+		}
+		return href;
+	}
+
+	private computeCustomStyleSheetIncludes(uri: Uri): string[] {
+		const styles = vscode.workspace.getConfiguration('markdown')['styles'];
+		if (styles && Array.isArray(styles)) {
+			return styles.map((style) => {
+				return `<link rel="stylesheet" href="${this.fixHref(uri, style)}" type="text/css" media="screen">`;
+			});
+		}
+		return [];
+	}
+
+	public provideTextDocumentContent(uri: Uri): Thenable<string> {
+
+		return vscode.workspace.openTextDocument(Uri.parse(uri.query)).then(document => {
+			const head = [].concat(
+				'<!DOCTYPE html>',
+				'<html>',
+				'<head>',
+				'<meta http-equiv="Content-type" content="text/html;charset=UTF-8">',
+				`<link rel="stylesheet" type="text/css" href="${this.getMediaPath('markdown.css')}" >`,
+				`<link rel="stylesheet" type="text/css" href="${this.getMediaPath('tomorrow.css')}" >`,
+				this.computeCustomStyleSheetIncludes(uri),
+				'</head>',
+				'<body>'
+			).join('\n');
+
+			const body = md.render(document.getText());
+
+			const tail = [
+				'</body>',
+				'</html>'
+			].join('\n');
+
+			return head + body + tail;
+		});
+	}
+
+	get onDidChange(): Event<Uri> {
+		return this._onDidChange.event;
+	}
+
+	public update(uri: Uri) {
+		this._onDidChange.fire(uri);
```

Comment:
Have you considered a timout here?  Spamming markdown-it may not always be a good idea.  This is more true when you add more plugins like `highlight` and when I was testing for math support `katex`.  I simply did something like:

```
            if (!this._waiting) {
                this._waiting = true;
                setTimeout(() => {
                    this._waiting = false;
                    this._onDidChange.fire(uri);
                }, 300);
            }
```
The timeout should probably be modifiable by the user, maybe even default to 0 with a note that for those who may experience problems.